runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/JStormStateInternals.java

```diff
@@ -0,0 +1,831 @@
+            throw new RuntimeException(e.getCause());
+          }
+        }
+
+        @Override
+        public boolean hasNext() {
+          return cursor < size;
+        }
+
+        @Override
+        public T next() {
+          if (cursor >= size) {
+            throw new NoSuchElementException();
+          }
+
+          T value = null;
+          try {
+            value = kvState.get(getComposedKey(cursor));
+          } catch (IOException e) {
+            LOG.error("Failed to read composed key-[{}]", getComposedKey(cursor));
+          }
+          cursor++;
+          return value;
+        }
+
+        @Override
+        public void remove() {
+          throw new UnsupportedOperationException();
+        }
+      }
+
+      BagStateIterable() {
+
+      }
+
+      @Override
+      public Iterator<T> iterator() {
+        return new BagStateIterator();
+      }
+
+      @Override
+      public String toString() {
+        return String.format("BagStateIterable: composedKey=%s", getComposedKey());
+      }
+    }
+  }
+
+  /**
+   * JStorm implementation of {@link CombiningState}.
+   */
+  private static class JStormCombiningState<InputT, AccumT, OutputT>
+      implements CombiningState<InputT, AccumT, OutputT> {
+
+    private final String id;
+    private final StateSpec<CombiningState<InputT, AccumT, OutputT>> spec;
+    private final StateNamespace namespace;
+    @Nullable
+    private final BagState<AccumT> accumBagState;
+    private final Combine.CombineFn<InputT, AccumT, OutputT> combineFn;
+
+    JStormCombiningState(
+        String id,
+        StateSpec<CombiningState<InputT, AccumT, OutputT>> spec,
+        StateNamespace namespace,
+        BagState<AccumT> accumBagState,
+        Combine.CombineFn<InputT, AccumT, OutputT> combineFn) {
+      this.id = id;
+      this.spec = spec;
+      this.namespace = namespace;
+      this.accumBagState = checkNotNull(accumBagState, "accumBagState");
+      this.combineFn = checkNotNull(combineFn, "combineFn");
+    }
+
+    @Override
+    public AccumT getAccum() {
+      // TODO: replacing the accumBagState with the merged accum.
+      return combineFn.mergeAccumulators(accumBagState.read());
+    }
+
+    @Override
+    public void addAccum(AccumT accumT) {
+      accumBagState.add(accumT);
+    }
+
+    @Override
+    public AccumT mergeAccumulators(Iterable<AccumT> iterable) {
+      return combineFn.mergeAccumulators(iterable);
+    }
+
+    @Override
+    public void add(InputT input) {
+      accumBagState.add(
+          combineFn.addInput(combineFn.createAccumulator(), input));
+    }
+
+    @Override
+    public ReadableState<Boolean> isEmpty() {
+      return accumBagState.isEmpty();
+    }
+
+    @Override
+    public OutputT read() {
+      return combineFn.extractOutput(
+          combineFn.mergeAccumulators(accumBagState.read()));
+    }
+
+    @Override
+    public CombiningState<InputT, AccumT, OutputT> readLater() {
+      // TODO: support prefetch.
```

Comment:
<!--new_thread; commit:13d5a9c5b6f60fd83d944812acd8500e37d17f48; resolved:0-->
I think it would be good to file JIRA for this (and other) TODOs. They may be lost in the large merge of the PR.